src/kudu/consensus/routing.cc

```diff
@@ -0,0 +1,162 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "kudu/consensus/routing.h"
+
+#include <glog/logging.h>
+
+#include "kudu/gutil/map-util.h"
+#include "kudu/util/status.h"
+
+using std::string;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+
+namespace kudu {
+namespace consensus {
+
+Status RoutingTable::Init(const RaftConfigPB& config,
+                          const std::string& leader_uuid) {
+  unordered_map<string, Node*> index;
+  unordered_map<string, unique_ptr<Node>> trees;
+
+  RETURN_NOT_OK(ConstructForest(config, &index, &trees));
+  RETURN_NOT_OK(MergeTrees(leader_uuid, index, &trees));
+  ConstructNextHopIndicesRec(trees.begin()->second.get());
+
+  index_ = std::move(index);
+  topology_root_ = std::move(trees.begin()->second);
+
+  return Status::OK();
+}
+
+Status RoutingTable::ConstructForest(
+    const RaftConfigPB& config,
+    std::unordered_map<std::string, Node*>* index,
+    std::unordered_map<std::string, std::unique_ptr<Node>>* trees) {
+
+  // Construct the peers.
+  for (const RaftPeerPB& peer : config.peers()) {
```

Comment:
the routing table is being built in memory it seems, and what it primarily depends on is proxy_from field.
It should be realtively easy to change peer.proxy_from()  to something like proxy_map_config.proxy_from(peer),
and with the same outcome as this code assumes.

Reply:
This is done, along with a bunch of other changes we discussed offline that are detailed in the individual commit messages in this stack.